File: src/Sample/Plumix.Sample/Demos/Cupertino/CupertinoLocalizationsDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia.Media;
using Plumix.Cupertino;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source (reference): dart_sample/lib/demos/cupertino/cupertino_localizations_demo_page.dart
// (exact sample parity)

public sealed class CupertinoLocalizationsDemoPage : StatefulWidget
{
    public override State CreateState() => new CupertinoLocalizationsDemoPageState();
}

internal sealed class CupertinoLocalizationsDemoPageState : State
{
    private static readonly IReadOnlyList<(string Label, Locale Locale)> Locales =
    [
        ("English", new Locale("en")),
        ("Français", new Locale("fr")),
        ("Русский", new Locale("ru")),
        ("中文", new Locale("zh")),
        ("العربية", new Locale("ar")),
    ];

    private static readonly DateTime SampleDate = new(2019, 3, 25);

    private int _selected;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12.0,
            children:
            [
                new Text("Cupertino localizations", fontSize: 20.0, color: Colors.Black),
                new Text(
                    "GlobalCupertinoLocalizations.Delegates translates every Cupertino string and "
                    + "formats dates with the locale's own calendar data.",
                    fontSize: 14.0,
                    color: Color.Parse("#8A000000")),
                new Wrap(spacing: 8.0, runSpacing: 8.0, children: BuildLocaleButtons()),
                new Localizations(
                    locale: Locales[_selected].Locale,
                    delegates: GlobalCupertinoLocalizations.Delegates,
                    child: new Builder(BuildTranslations)),
            ]);
    }

    private IReadOnlyList<Widget> BuildLocaleButtons()
    {
        var buttons = new List<Widget>();
        for (int index = 0; index < Locales.Count; index++)
        {
            int target = index;
            bool selected = index == _selected;
            buttons.Add(
                new GestureDetector(
                    behavior: HitTestBehavior.Opaque,
                    onTap: () => SetState(() => _selected = target),
                    child: new Container(
                        padding: EdgeInsets.Symmetric(horizontal: 14.0, vertical: 8.0),
                        decoration: new BoxDecoration(
                            Color: selected ? Color.Parse("#FF007AFF") : Color.Parse("#FFF2F2F7"),
                            BorderRadius: BorderRadius.Circular(10.0)),
                        child: new Text(
                            Locales[target].Label,
                            fontSize: 14.0,
                            color: selected ? Colors.White : Colors.Black))));
        }

        return buttons;
    }

    private static Widget BuildTranslations(BuildContext context)
    {
        CupertinoLocalizations localizations = CupertinoLocalizations.Of(context);
        WidgetsLocalizations widgets = WidgetsLocalizations.Of(context);

        return new Container(
            padding: EdgeInsets.All(16.0),
            decoration: new BoxDecoration(
                Color: Color.Parse("#FFF2F2F7"),
                BorderRadius: BorderRadius.Circular(12.0)),
            child: new Column(
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 6.0,
                children:
                [
                    BuildRow("textDirection", widgets.TextDirection.ToString()),
                    BuildRow("alertDialogLabel", localizations.AlertDialogLabel),
                    BuildRow(
                        "toolbar",
                        string.Join(
                            " · ",
                            localizations.CutButtonLabel,
                            localizations.CopyButtonLabel,
                            localizations.PasteButtonLabel,
                            localizations.SelectAllButtonLabel)),
                    BuildRow("datePickerMediumDate", localizations.DatePickerMediumDate(SampleDate)),
                    BuildRow(
                        "datePickerMonth / standalone",
                        $"{localizations.DatePickerMonth(5)} / {localizations.DatePickerStandaloneMonth(5)}"),
                    BuildRow(
                        "datePickerDayOfMonth",
                        $"{localizations.DatePickerDayOfMonth(1)} · {localizations.DatePickerDayOfMonth(1, 2)}"),
                    BuildRow(
                        "datePickerHourSemanticsLabel",
                        $"{localizations.DatePickerHourSemanticsLabel(1)} · "
                        + $"{localizations.DatePickerHourSemanticsLabel(12)}"),
                    BuildRow(
                        "timerPicker",
                        $"{localizations.TimerPickerMinute(10)} "
                        + $"{localizations.TimerPickerMinuteLabel(10)} · "
                        + $"{localizations.TimerPickerHour(1)} {localizations.TimerPickerHourLabel(1)}"),
                    BuildRow("tabSemanticsLabel", localizations.TabSemanticsLabel(1, 2)),
                    BuildRow("reorderItemUp", widgets.ReorderItemUp),
                ]));
    }

    private static Widget BuildRow(string label, string value)
    {
        return new Row(
            crossAxisAlignment: CrossAxisAlignment.Start,
            spacing: 12.0,
            children:
            [
                new SizedBox(
                    width: 220.0,
                    child: new Text(label, fontSize: 13.0, color: Color.Parse("#8A000000"))),
                new Expanded(child: new Text(value, fontSize: 13.0, color: Colors.Black)),
            ]);
    }
}

File: dart_sample/lib/demos/cupertino/cupertino_localizations_demo_page.dart
import 'package:cupertino_ui/cupertino_ui.dart';
import 'package:material_ui/material_ui.dart';

class CupertinoLocalizationsDemoPage extends StatefulWidget {
  const CupertinoLocalizationsDemoPage({super.key});

  @override
  State<CupertinoLocalizationsDemoPage> createState() =>
      _CupertinoLocalizationsDemoPageState();
}

class _CupertinoLocalizationsDemoPageState
    extends State<CupertinoLocalizationsDemoPage> {
  static const List<(String, Locale)> _locales = <(String, Locale)>[
    ('English', Locale('en')),
    ('Français', Locale('fr')),
    ('Русский', Locale('ru')),
    ('中文', Locale('zh')),
    ('العربية', Locale('ar')),
  ];

  static final DateTime _sampleDate = DateTime(2019, 3, 25);

  int _selected = 0;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'Cupertino localizations',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'GlobalCupertinoLocalizations.delegates translates every Cupertino '
          "string and formats dates with the locale's own calendar data.",
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Wrap(spacing: 8, runSpacing: 8, children: _buildLocaleButtons()),
        Localizations(
          locale: _locales[_selected].$2,
          delegates: GlobalCupertinoLocalizations.delegates,
          child: Builder(builder: _buildTranslations),
        ),
      ],
    );
  }

  List<Widget> _buildLocaleButtons() {
    return <Widget>[
      for (int index = 0; index < _locales.length; index++)
        GestureDetector(
          behavior: HitTestBehavior.opaque,
          onTap: () => setState(() => _selected = index),
          child: Container(
            padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 8),
            decoration: BoxDecoration(
              color: index == _selected
                  ? const Color(0xFF007AFF)
                  : const Color(0xFFF2F2F7),
              borderRadius: BorderRadius.circular(10),
            ),
            child: Text(
              _locales[index].$1,
              style: TextStyle(
                fontSize: 14,
                color: index == _selected ? Colors.white : Colors.black,
              ),
            ),
          ),
        ),
    ];
  }

  Widget _buildTranslations(BuildContext context) {
    final CupertinoLocalizations localizations = CupertinoLocalizations.of(
      context,
    );
    final WidgetsLocalizations widgets = WidgetsLocalizations.of(context);

    return Container(
      padding: const EdgeInsets.all(16),
      decoration: BoxDecoration(
        color: const Color(0xFFF2F2F7),
        borderRadius: BorderRadius.circular(12),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        spacing: 6,
        children: <Widget>[
          _buildRow('textDirection', widgets.textDirection.name),
          _buildRow('alertDialogLabel', localizations.alertDialogLabel),
          _buildRow(
            'toolbar',
            <String>[
              localizations.cutButtonLabel,
              localizations.copyButtonLabel,
              localizations.pasteButtonLabel,
              localizations.selectAllButtonLabel,
            ].join(' · '),
          ),
          _buildRow(
            'datePickerMediumDate',
            localizations.datePickerMediumDate(_sampleDate),
          ),
          _buildRow(
            'datePickerMonth / standalone',
            '${localizations.datePickerMonth(5)} / '
                '${localizations.datePickerStandaloneMonth(5)}',
          ),
          _buildRow(
            'datePickerDayOfMonth',
            '${localizations.datePickerDayOfMonth(1)} · '
                '${localizations.datePickerDayOfMonth(1, 2)}',
          ),
          _buildRow(
            'datePickerHourSemanticsLabel',
            '${localizations.datePickerHourSemanticsLabel(1)} · '
                '${localizations.datePickerHourSemanticsLabel(12)}',
          ),
          _buildRow(
            'timerPicker',
            '${localizations.timerPickerMinute(10)} '
                '${localizations.timerPickerMinuteLabel(10)} · '
                '${localizations.timerPickerHour(1)} '
                '${localizations.timerPickerHourLabel(1)}',
          ),
          _buildRow(
            'tabSemanticsLabel',
            localizations.tabSemanticsLabel(tabIndex: 1, tabCount: 2),
          ),
          _buildRow('reorderItemUp', widgets.reorderItemUp),
        ],
      ),
    );
  }

  Widget _buildRow(String label, String value) {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      spacing: 12,
      children: <Widget>[
        SizedBox(
          width: 220,
          child: Text(
            label,
            style: const TextStyle(fontSize: 13, color: Colors.black54),
          ),
        ),
        Expanded(
          child: Text(
            value,
            style: const TextStyle(fontSize: 13, color: Colors.black),
          ),
        ),
      ],
    );
  }
}
